railties/lib/rails/tasks/yarn.rake

```diff
@@ -10,17 +10,21 @@ namespace :yarn do
     end
 
     yarn_flags =
-      if `#{Rails.root}/bin/yarn --version`.start_with?("1")
+      if `"#{Rails.root}/bin/yarn" --version`.start_with?("1")
         "--no-progress --frozen-lockfile"
       else
         "--immutable"
       end
 
-    system({ "NODE_ENV" => node_env }, "#{Rails.root}/bin/yarn install #{yarn_flags}")
+    system({ "NODE_ENV" => node_env }, "\"#{Rails.root}/bin/yarn\" install #{yarn_flags}")
+  rescue Errno::ENOENT
+    $stderr.puts "bin/yarn was not found."
+    $stderr.puts "Please run `bundle exec rails app:update:bin` to create it."
+    exit 1
   end
 end
 
 # Run Yarn prior to Sprockets assets precompilation, so dependencies are available for use.
-if Rake::Task.task_defined?("assets:precompile")
+if Rake::Task.task_defined?("assets:precompile") && File.exist?(Rails.root.join("bin", "yarn"))
```

Comment:
After further comments in #40795 it looks like `yarn:install` is *always* run, even if javascript was skipped in the Rails project and `rails app:update:bin` therefore does *not* generate `bin/yarn`.

I added this check to make sure `bin/yarn` exists. It is the same as done in the app updater:

https://github.com/rails/rails/blob/b087511bc49097a6dc2afe0256bb6d39ebb3a4ab/railties/lib/rails/app_updater.rb#L24

Now it should really only be run automatically before compiling assets if `bin/yarn` is present.